### ISSUE TITLE
False positive in RedundantSuspendModifier with kotlin 1.9.0

### ISSUE THREAD

**Issue by josephlbarnett** [detekt/detekt#6276]
## Expected Behavior
detekt should pass regardless of kotlin version

## Observed Behavior
detekt passes in kotlin 1.8.22 but with kotlin version 1.9.0, detekt shows RedundantSuspendModifier in https://github.com/trib3/leakycauldron/blob/main/server/src/main/kotlin/com/trib3/server/coroutine/CoroutineInvocationHandler.kt#L34, but this suspend function calls suspendCoroutineUninterceptedOrReturn so needs to be suspend

## Steps to Reproduce

https://github.com/trib3/leakycauldron/blob/main/server/src/main/kotlin/com/trib3/server/coroutine/CoroutineInvocationHandler.kt but bump version.kotlin in parent-pom/pom.xml to 1.9.0

## Your Environment

* Version of detekt used: 1.22.0
* Version of Gradle used (if applicable):
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: ubuntu 23.04
* Link to your project (if it's a public repository): https://github.com/trib3/leakycauldron

**Comment by nacyolsa** [comment]
I receive false positives for RedundantSuspendModifier also in 1.8.21.

**Comment by cortinico** [comment]
> I receive false positives for RedundantSuspendModifier also in Kotlin 1.8.21 and Detekt 1.23.0.

@nacyolsa can we get a snippets that reproduces?
I believe this is not an Kotlin 1.9.0 issue but it's just a false positive that should be fixed 👍

**Comment by nacyolsa** [comment]
@cortinico One of many examples:

```kotlin
class ClientABC(
    val httpClient: HttpClient = HttpClient {}
) {

    suspend inline fun <reified T> get() =
        Either.catch {
            httpClient.use {
                it.get {
                    url {
                        path("")
                    }
                }
            }.body<T>()
        }.mapLeft {}
}
```

**Comment by t-kameyama** [comment]
@josephlbarnett Please try detekt 1.23.0

**Comment by t-kameyama** [comment]
Simple code to reproduce false positive:
```kotlin
suspend fun test(foo: Foo) {
    foo.use {
        it.get()
    }
}

class Foo: AutoCloseable {
    @Suppress("RedundantSuspendModifier")
    suspend fun get() = Unit
    override fun close() = Unit
}
```

**Comment by atulgpt** [comment]
Hi @t-kameyama why the above code is false positive as `suspend fun get() = Unit` can be written as `fun get() = Unit`

Won't compliant code be as below
```
suspend fun test(foo: Foo) {
    foo.use {
        it.get()
    }
}

class Foo: AutoCloseable {
    fun get() = Unit
    override fun close() = Unit
}
```

**Comment by josephlbarnett** [comment]
> @josephlbarnett Please try detekt 1.23.0

it does appear to work with detekt 1.23.1 and kotlin 1.9.0 (but #6129 is still an issue with those versions)

**Comment by cortinico** [comment]
> it does appear to work with detekt 1.23.1 and kotlin 1.9.0 (but #6129 is still an issue with those versions)

Closing then 👍 we can follow up on the other issue
